File: src/domain/note-links.ts
import { decodeMediaHref, noteDirectory, resolveWorkspaceFilePath } from "./paths";

export type NoteLinkKind = "wiki" | "markdown";

export type ExtractedNoteLink = {
  targetRef: string;
  heading: string;
  displayText: string;
  kind: NoteLinkKind;
};

export type NoteLinkIdentity = {
  relativePath: string;
  title: string;
};

export type NoteGraphNode = {
  relativePath: string;
  title: string;
  folder: string;
};

export type NoteGraphEdge = {
  sourcePath: string;
  targetPath: string | null;
  targetRef: string;
  displayText: string;
  heading: string;
  kind: NoteLinkKind;
};

export type NoteGraph = {
  nodes: NoteGraphNode[];
  edges: NoteGraphEdge[];
};

export type NoteLinkItem = {
  sourcePath: string;
  sourceTitle: string;
  targetPath: string | null;
  targetTitle: string | null;
  targetRef: string;
  displayText: string;
  heading: string;
  kind: NoteLinkKind;
};

export type NoteRefs = {
  outgoing: NoteLinkItem[];
  incoming: NoteLinkItem[];
  unresolved: NoteLinkItem[];
};

const MEDIA_EXT = /\.(png|jpe?g|gif|webp|bmp|avif|svg|ico|pdf)$/i;
const NOTE_EXT = /\.(md|mdx)$/i;
const WIKI_RE = /\[\[([^\[\]]+)\]\]/g;
const MD_LINK_RE = /\[([^\]]*)\]\(([^)]+)\)/g;
const MD_IMAGE_RE = /!\[[^\]]*\]\([^)]+\)/g;
const WIKI_EMBED_RE = /!\[\[[^\[\]]+\]\]/g;

export function folderOf(relativePath: string) {
  return noteDirectory(relativePath);
}

export function noteStem(relativePath: string) {
  const fileName = relativePath.split("/").pop() || relativePath;
  return fileName.replace(NOTE_EXT, "");
}

export function extractNoteLinks(content: string): ExtractedNoteLink[] {
  const visible = stripInlineCode(visibleMarkdown(stripFrontmatter(content)))
    .replace(WIKI_EMBED_RE, " ")
    .replace(MD_IMAGE_RE, " ");
  const links: ExtractedNoteLink[] = [];
  const seen = new Set<string>();

  WIKI_RE.lastIndex = 0;
  for (const match of visible.matchAll(WIKI_RE)) {
    const index = match.index ?? 0;
    if (index > 0 && visible[index - 1] === "!") continue;
    const parsed = parseWikiInner(match[1] || "");
    if (!parsed || !rememberLink(seen, parsed)) continue;
    links.push(parsed);
  }

  MD_LINK_RE.lastIndex = 0;
  for (const match of visible.matchAll(MD_LINK_RE)) {
    const index = match.index ?? 0;
    if (index > 0 && visible[index - 1] === "!") continue;
    const parsed = parseMarkdownLink(match[1] || "", match[2] || "");
    if (!parsed || !rememberLink(seen, parsed)) continue;
    links.push(parsed);
  }

  return links;
}

export function resolveNoteRef(
  targetRef: string,
  sourcePath: string,
  notes: NoteLinkIdentity[],
): string | undefined {
  const normalized = normalizeTargetRef(targetRef);
  if (!normalized) return undefined;

  const byPath = new Map<string, string>();
  const byStem = new Map<string, string[]>();
  const byTitle = new Map<string, string[]>();
  for (const note of notes) {
    const path = note.relativePath.replace(/\\/g, "/");
    byPath.set(path.toLowerCase(), path);
    const withoutExt = path.replace(NOTE_EXT, "");
    byPath.set(withoutExt.toLowerCase(), path);
    pushIndex(byStem, noteStem(path).toLowerCase(), path);
    const title = note.title.trim().toLowerCase();
    if (title) pushIndex(byTitle, title, path);
  }

  const sourceDir = folderOf(sourcePath);
  const looksLikePath = normalized.includes("/") || /^\.\.?(?:\/|$)/.test(normalized) || NOTE_EXT.test(normalized);

  if (looksLikePath) {
    const relative = joinRelative(sourceDir, normalized);
    const fromRelative = matchPath(relative, byPath);
    if (fromRelative) return fromRelative;
    const fromRoot = matchPath(normalized.replace(/^\.\//, ""), byPath);
    if (fromRoot) return fromRoot;
  }

  const exact = matchPath(normalized, byPath);
  if (exact) return exact;

  const stemHit = pickUnique(byStem.get(normalized.toLowerCase()) || [], sourceDir);
  if (stemHit) return stemHit;

  const titleHit = pickUnique(byTitle.get(normalized.toLowerCase()) || [], sourceDir);
  if (titleHit) return titleHit;

  return undefined;
}

export function buildNoteGraph(
  notes: Array<{ relativePath: string; title: string; content: string }>,
): NoteGraph {
  const identities = notes.map((note) => ({
    relativePath: note.relativePath.replace(/\\/g, "/"),
    title: note.title,
  }));
  const nodes = identities.map((note) => ({
    relativePath: note.relativePath,
    title: note.title,
    folder: folderOf(note.relativePath),
  }));
  const edges: NoteGraphEdge[] = [];
  const seen = new Set<string>();
  for (const note of notes) {
    const sourcePath = note.relativePath.replace(/\\/g, "/");
    for (const link of extractNoteLinks(note.content)) {
      const targetPath = resolveNoteRef(link.targetRef, sourcePath, identities) ?? null;
      const key = `${sourcePath}\0${targetPath ?? ""}\0${link.targetRef}\0${link.kind}\0${link.heading}`;
      if (seen.has(key)) continue;
      seen.add(key);
      edges.push({
        sourcePath,
        targetPath,
        targetRef: link.targetRef,
        displayText: link.displayText,
        heading: link.heading,
        kind: link.kind,
      });
    }
  }
  return { nodes, edges };
}

export function noteRefsFromGraph(
  graph: NoteGraph,
  relativePath: string,
  liveOutgoing?: ExtractedNoteLink[],
): NoteRefs {
  const path = relativePath.replace(/\\/g, "/");
  const titles = new Map(graph.nodes.map((node) => [node.relativePath, node.title]));
  const sourceTitle = titles.get(path) || noteStem(path);
  const outgoingSource =
    liveOutgoing ??
    graph.edges
      .filter((edge) => edge.sourcePath === path)
      .map((edge) => ({
        targetRef: edge.targetRef,
        heading: edge.heading,
        displayText: edge.displayText,
        kind: edge.kind,
      }));
  const identities = graph.nodes.map((node) => ({
    relativePath: node.relativePath,
    title: node.title,
  }));
  const outgoing: NoteLinkItem[] = outgoingSource.map((link) => {
    const targetPath = resolveNoteRef(link.targetRef, path, identities) ?? null;
    return {
      sourcePath: path,
      sourceTitle,
      targetPath,
      targetTitle: targetPath ? titles.get(targetPath) || null : null,
      targetRef: link.targetRef,
      displayText: link.displayText,
      heading: link.heading,
      kind: link.kind,
    };
  });
  const incoming = graph.edges
    .filter((edge) => edge.targetPath === path && edge.sourcePath !== path)
    .map((edge) => ({
      sourcePath: edge.sourcePath,
      sourceTitle: titles.get(edge.sourcePath) || noteStem(edge.sourcePath),
      targetPath: path,
      targetTitle: sourceTitle,
      targetRef: edge.targetRef,
      displayText: edge.displayText,
      heading: edge.heading,
      kind: edge.kind,
    }));
  return {
    outgoing,
    incoming,
    unresolved: outgoing.filter((item) => !item.targetPath),
  };
}

export function isNoteMarkdownHref(href: string) {
  const path = stripHash(decodeMediaHref(href)).replace(/\\/g, "/");
  if (!path || isExternalHref(href) || isMediaRef(path)) return false;
  return NOTE_EXT.test(path);
}

export function splitHash(value: string): { path: string; heading: string } {
  const trimmed = value.trim();
  const hash = trimmed.indexOf("#");
  if (hash < 0) return { path: trimmed, heading: "" };
  return {
    path: trimmed.slice(0, hash).trim(),
    heading: trimmed.slice(hash + 1).trim(),
  };
}

function parseWikiInner(inner: string): ExtractedNoteLink | null {
  const trimmed = inner.trim();
  if (!trimmed) return null;
  const pipe = trimmed.indexOf("|");
  const targetPart = (pipe < 0 ? trimmed : trimmed.slice(0, pipe)).trim();
  const alias = (pipe < 0 ? "" : trimmed.slice(pipe + 1)).trim();
  const { path, heading } = splitHash(targetPart);
  const targetRef = normalizeTargetRef(path);
  if (!targetRef || isMediaRef(targetRef)) return null;
  return {
    targetRef,
    heading,
    displayText: alias || displayFromRef(targetRef, heading),
    kind: "wiki",
  };
}

function parseMarkdownLink(label: string, href: string): ExtractedNoteLink | null {
  const raw = href.trim().replace(/^<|>$/g, "");
  if (!isNoteMarkdownHref(raw)) return null;
  const { path, heading } = splitHash(decodeMediaHref(raw));
  const targetRef = normalizeTargetRef(path);
  if (!targetRef) return null;
  const display = label.trim() || displayFromRef(targetRef, heading);
  return {
    targetRef,
    heading,
    displayText: display,
    kind: "markdown",
  };
}

function rememberLink(seen: Set<string>, link: ExtractedNoteLink) {
  const key = `${link.kind}\0${link.targetRef}\0${link.heading}`;
  if (seen.has(key)) return false;
  seen.add(key);
  return true;
}

function stripFrontmatter(content: string) {
  return content.replace(/^\uFEFF/, "").replace(/^---\r?\n[\s\S]*?\r?\n---\r?\n?/, "");
}

function visibleMarkdown(content: string) {
  let inFence = false;
  const lines: string[] = [];
  for (const line of content.split("\n")) {
    if (/^\s*(```|~~~)/.test(line)) {
      inFence = !inFence;
      lines.push("");
      continue;
    }
    lines.push(inFence ? "" : line);
  }
  return lines.join("\n");
}

function stripInlineCode(content: string) {
  const chars = [...content];
  const out: string[] = [];
  let index = 0;
  while (index < chars.length) {
    if (chars[index] !== "`") {
      out.push(chars[index]);
      index += 1;
      continue;
    }
    let ticks = 0;
    while (index + ticks < chars.length && chars[index + ticks] === "`") ticks += 1;
    let cursor = index + ticks;
    let end = chars.length;
    while (cursor < chars.length) {
      if (chars[cursor] !== "`") {
        cursor += 1;
        continue;
      }
      let close = 0;
      while (cursor + close < chars.length && chars[cursor + close] === "`") close += 1;
      if (close === ticks) {
        end = cursor + close;
        break;
      }
      cursor += close;
    }
    for (let fill = index; fill < end; fill += 1) out.push(" ");
    index = end;
  }
  return out.join("");
}

function normalizeTargetRef(value: string) {
  return value.replace(/\\/g, "/").replace(/^\.\//, "").replace(/\/+$/, "").trim();
}

function displayFromRef(targetRef: string, heading: string) {
  const name = noteStem(targetRef.split("/").pop() || targetRef);
  return heading ? `${name}#${heading}` : name;
}

function isExternalHref(href: string) {
  return /^(https?:|mailto:|data:|javascript:|ftp:|tel:|#)/i.test(href.trim());
}

function isMediaRef(path: string) {
  const clean = stripHash(path);
  return MEDIA_EXT.test(clean) || /(^|\/)attachments\//i.test(clean);
}

function stripHash(value: string) {
  const hash = value.indexOf("#");
  return (hash < 0 ? value : value.slice(0, hash)).trim();
}

function pushIndex(map: Map<string, string[]>, key: string, path: string) {
  const list = map.get(key);
  if (list) {
    if (!list.includes(path)) list.push(path);
  } else {
    map.set(key, [path]);
  }
}

function matchPath(candidate: string, byPath: Map<string, string>) {
  const direct = byPath.get(candidate.toLowerCase());
  if (direct) return direct;
  if (NOTE_EXT.test(candidate)) return undefined;
  return byPath.get(`${candidate}.md`.toLowerCase()) || byPath.get(`${candidate}.mdx`.toLowerCase());
}

function joinRelative(sourceDir: string, target: string) {
  const joined = resolveWorkspaceFilePath(sourceDir || ".", target).replace(/\\/g, "/");
  return joined.replace(/^\.\//, "").replace(/^\/+/, "");
}

function pickUnique(candidates: string[], sourceDir: string) {
  if (candidates.length === 1) return candidates[0];
  if (!candidates.length) return undefined;
  const sameFolder = candidates.filter((path) => folderOf(path) === sourceDir);
  if (sameFolder.length === 1) return sameFolder[0];
  const pool = sameFolder.length ? sameFolder : candidates;
  const shortest = Math.min(...pool.map((path) => path.length));
  const short = pool.filter((path) => path.length === shortest);
  return short.length === 1 ? short[0] : undefined;
}

File: src-tauri/src/domain/note_links.rs
use crate::domain::folder_of;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::OnceLock;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NoteLinkKind {
    Wiki,
    Markdown,
}

impl NoteLinkKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Wiki => "wiki",
            Self::Markdown => "markdown",
        }
    }

    pub fn parse(value: &str) -> Self {
        if value.eq_ignore_ascii_case("markdown") {
            Self::Markdown
        } else {
            Self::Wiki
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RawNoteLink {
    pub target_ref: String,
    pub heading: String,
    pub display_text: String,
    pub kind: NoteLinkKind,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NoteLinkIdentity {
    pub relative_path: String,
    pub title: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NoteGraphNode {
    pub relative_path: String,
    pub title: String,
    pub folder: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NoteGraphEdge {
    pub source_path: String,
    pub target_path: Option<String>,
    pub target_ref: String,
    pub display_text: String,
    pub heading: String,
    pub kind: NoteLinkKind,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct NoteGraph {
    pub nodes: Vec<NoteGraphNode>,
    pub edges: Vec<NoteGraphEdge>,
}

pub fn extract_note_links(content: &str) -> Vec<RawNoteLink> {
    let content = content.strip_prefix('\u{feff}').unwrap_or(content);
    let stripped = strip_frontmatter(content);
    let visible = strip_inline_code(&visible_markdown(&stripped));
    let visible = wiki_embed_re().replace_all(&visible, " ");
    let visible = md_image_re().replace_all(&visible, " ");

    let mut links = Vec::new();
    let mut seen = HashSet::new();

    for captured in wiki_re().captures_iter(&visible) {
        let Some(full) = captured.get(0) else {
            continue;
        };
        if full.start() > 0 && visible.as_bytes().get(full.start() - 1) == Some(&b'!') {
            continue;
        }
        if let Some(parsed) =
            parse_wiki_inner(captured.get(1).map(|item| item.as_str()).unwrap_or(""))
        {
            if remember_link(&mut seen, &parsed) {
                links.push(parsed);
            }
        }
    }

    for captured in md_link_re().captures_iter(&visible) {
        let Some(full) = captured.get(0) else {
            continue;
        };
        if full.start() > 0 && visible.as_bytes().get(full.start() - 1) == Some(&b'!') {
            continue;
        }
        let label = captured.get(1).map(|item| item.as_str()).unwrap_or("");
        let href = captured.get(2).map(|item| item.as_str()).unwrap_or("");
        if let Some(parsed) = parse_markdown_link(label, href) {
            if remember_link(&mut seen, &parsed) {
                links.push(parsed);
            }
        }
    }

    links
}

pub fn resolve_note_ref(
    target_ref: &str,
    source_path: &str,
    notes: &[NoteLinkIdentity],
) -> Option<String> {
    let normalized = normalize_target_ref(target_ref);
    if normalized.is_empty() {
        return None;
    }

    let mut by_path: HashMap<String, String> = HashMap::new();
    let mut by_stem: HashMap<String, Vec<String>> = HashMap::new();
    let mut by_title: HashMap<String, Vec<String>> = HashMap::new();
    for note in notes {
        let path = note.relative_path.replace('\\', "/");
        by_path.insert(path.to_ascii_lowercase(), path.clone());
        let without_ext = strip_note_ext(&path);
        by_path.insert(without_ext.to_ascii_lowercase(), path.clone());
        push_index(
            &mut by_stem,
            note_stem(&path).to_ascii_lowercase(),
            path.clone(),
        );
        let title = note.title.trim().to_ascii_lowercase();
        if !title.is_empty() {
            push_index(&mut by_title, title, path);
        }
    }

    let source_dir = folder_of(source_path);
    let looks_like_path = normalized.contains('/')
        || normalized == "."
        || normalized.starts_with("./")
        || normalized.starts_with("..")
        || has_note_ext(&normalized);

    if looks_like_path {
        let relative = join_relative(&source_dir, &normalized);
        if let Some(found) = match_path(&relative, &by_path) {
            return Some(found);
        }
        let from_root = normalized.trim_start_matches("./");
        if let Some(found) = match_path(from_root, &by_path) {
            return Some(found);
        }
    }

    if let Some(found) = match_path(&normalized, &by_path) {
        return Some(found);
    }
    if let Some(found) = pick_unique(by_stem.get(&normalized.to_ascii_lowercase()), &source_dir) {
        return Some(found);
    }
    pick_unique(by_title.get(&normalized.to_ascii_lowercase()), &source_dir)
}

fn parse_wiki_inner(inner: &str) -> Option<RawNoteLink> {
    let trimmed = inner.trim();
    if trimmed.is_empty() {
        return None;
    }
    let (target_part, alias) = match trimmed.split_once('|') {
        Some((target, alias)) => (target.trim(), alias.trim()),
        None => (trimmed, ""),
    };
    let (path, heading) = split_hash(target_part);
    let target_ref = normalize_target_ref(&path);
    if target_ref.is_empty() || is_media_ref(&target_ref) {
        return None;
    }
    let display_text = if alias.is_empty() {
        display_from_ref(&target_ref, &heading)
    } else {
        alias.to_string()
    };
    Some(RawNoteLink {
        target_ref,
        heading,
        display_text,
        kind: NoteLinkKind::Wiki,
    })
}

fn parse_markdown_link(label: &str, href: &str) -> Option<RawNoteLink> {
    let raw = href.trim().trim_start_matches('<').trim_end_matches('>');
    if !is_note_markdown_href(raw) {
        return None;
    }
    let decoded = percent_decode(raw);
    let (path, heading) = split_hash(&decoded);
    let target_ref = normalize_target_ref(&path);
    if target_ref.is_empty() {
        return None;
    }
    let display_text = {
        let label = label.trim();
        if label.is_empty() {
            display_from_ref(&target_ref, &heading)
        } else {
            label.to_string()
        }
    };
    Some(RawNoteLink {
        target_ref,
        heading,
        display_text,
        kind: NoteLinkKind::Markdown,
    })
}

fn is_note_markdown_href(href: &str) -> bool {
    let path = split_hash(&percent_decode(href)).0.replace('\\', "/");
    if path.is_empty() || is_external_href(href) || is_media_ref(&path) {
        return false;
    }
    has_note_ext(&path)
}

fn remember_link(seen: &mut HashSet<String>, link: &RawNoteLink) -> bool {
    let key = format!(
        "{}\0{}\0{}",
        link.kind.as_str(),
        link.target_ref,
        link.heading
    );
    seen.insert(key)
}

fn strip_frontmatter(content: &str) -> String {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    let pattern =
        PATTERN.get_or_init(|| Regex::new(r"^---\r?\n[\s\S]*?\r?\n---\r?\n?").expect("static"));
    pattern.replace(content, "").into_owned()
}

fn visible_markdown(content: &str) -> String {
    static FENCE: OnceLock<Regex> = OnceLock::new();
    let fence = FENCE.get_or_init(|| Regex::new(r"^\s*(```|~~~)").expect("static"));
    let mut in_fence = false;
    let mut lines = Vec::new();
    for line in content.split('\n') {
        if fence.is_match(line) {
            in_fence = !in_fence;
            lines.push(String::new());
            continue;
        }
        lines.push(if in_fence {
            String::new()
        } else {
            line.to_string()
        });
    }
    lines.join("\n")
}

fn strip_inline_code(content: &str) -> String {
    let chars: Vec<char> = content.chars().collect();
    let mut out = String::with_capacity(content.len());
    let mut index = 0;
    while index < chars.len() {
        if chars[index] != '`' {
            out.push(chars[index]);
            index += 1;
            continue;
        }
        let mut ticks = 0;
        while index + ticks < chars.len() && chars[index + ticks] == '`' {
            ticks += 1;
        }
        let mut cursor = index + ticks;
        let mut end = chars.len();
        while cursor < chars.len() {
            if chars[cursor] != '`' {
                cursor += 1;
                continue;
            }
            let mut close = 0;
            while cursor + close < chars.len() && chars[cursor + close] == '`' {
                close += 1;
            }
            if close == ticks {
                end = cursor + close;
                break;
            }
            cursor += close;
        }
        for _ in index..end {
            out.push(' ');
        }
        index = end;
    }
    out
}

fn split_hash(value: &str) -> (String, String) {
    let trimmed = value.trim();
    match trimmed.split_once('#') {
        Some((path, heading)) => (path.trim().to_string(), heading.trim().to_string()),
        None => (trimmed.to_string(), String::new()),
    }
}

fn normalize_target_ref(value: &str) -> String {
    value
        .replace('\\', "/")
        .trim()
        .trim_start_matches("./")
        .trim_end_matches('/')
        .to_string()
}

fn display_from_ref(target_ref: &str, heading: &str) -> String {
    let name = note_stem(target_ref.rsplit('/').next().unwrap_or(target_ref));
    if heading.is_empty() {
        name
    } else {
        format!("{name}#{heading}")
    }
}

fn is_external_href(href: &str) -> bool {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    let pattern = PATTERN.get_or_init(|| {
        Regex::new(r"^(https?:|mailto:|data:|javascript:|ftp:|tel:|#)").expect("static")
    });
    pattern.is_match(href.trim())
}

fn is_media_ref(path: &str) -> bool {
    static MEDIA: OnceLock<Regex> = OnceLock::new();
    static ATTACH: OnceLock<Regex> = OnceLock::new();
    let media = MEDIA.get_or_init(|| {
        Regex::new(r"(?i)\.(png|jpe?g|gif|webp|bmp|avif|svg|ico|pdf)$").expect("static")
    });
    let attach = ATTACH.get_or_init(|| Regex::new(r"(?i)(^|/)attachments/").expect("static"));
    let clean = split_hash(path).0;
    media.is_match(&clean) || attach.is_match(&clean)
}

fn has_note_ext(path: &str) -> bool {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    let pattern = PATTERN.get_or_init(|| Regex::new(r"(?i)\.(md|mdx)$").expect("static"));
    pattern.is_match(path)
}

fn strip_note_ext(path: &str) -> String {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    let pattern = PATTERN.get_or_init(|| Regex::new(r"(?i)\.(md|mdx)$").expect("static"));
    pattern.replace(path, "").into_owned()
}

fn note_stem(path: &str) -> String {
    strip_note_ext(path.rsplit('/').next().unwrap_or(path))
}

fn percent_decode(value: &str) -> String {
    let bytes = value.as_bytes();
    let mut out = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'%' && index + 2 < bytes.len() {
            let hex = &value[index + 1..index + 3];
            if let Ok(byte) = u8::from_str_radix(hex, 16) {
                out.push(byte);
                index += 3;
                continue;
            }
        }
        out.push(bytes[index]);
        index += 1;
    }
    String::from_utf8_lossy(&out).into_owned()
}

fn push_index(map: &mut HashMap<String, Vec<String>>, key: String, path: String) {
    let list = map.entry(key).or_default();
    if !list.iter().any(|item| item == &path) {
        list.push(path);
    }
}

fn match_path(candidate: &str, by_path: &HashMap<String, String>) -> Option<String> {
    if let Some(found) = by_path.get(&candidate.to_ascii_lowercase()) {
        return Some(found.clone());
    }
    if has_note_ext(candidate) {
        return None;
    }
    by_path
        .get(&format!("{candidate}.md").to_ascii_lowercase())
        .or_else(|| by_path.get(&format!("{candidate}.mdx").to_ascii_lowercase()))
        .cloned()
}

fn join_relative(source_dir: &str, target: &str) -> String {
    let mut parts: Vec<String> = if source_dir.is_empty() {
        Vec::new()
    } else {
        source_dir
            .split('/')
            .filter(|part| !part.is_empty())
            .map(ToOwned::to_owned)
            .collect()
    };
    for part in target.replace('\\', "/").split('/') {
        if part.is_empty() || part == "." {
            continue;
        }
        if part == ".." {
            parts.pop();
            continue;
        }
        parts.push(part.to_string());
    }
    parts.join("/")
}

fn pick_unique(candidates: Option<&Vec<String>>, source_dir: &str) -> Option<String> {
    let candidates = candidates?;
    if candidates.len() == 1 {
        return candidates.first().cloned();
    }
    if candidates.is_empty() {
        return None;
    }
    let same_folder: Vec<&String> = candidates
        .iter()
        .filter(|path| folder_of(path) == source_dir)
        .collect();
    let pool: Vec<&String> = if same_folder.len() == 1 {
        return Some(same_folder[0].clone());
    } else if !same_folder.is_empty() {
        same_folder
    } else {
        candidates.iter().collect()
    };
    let shortest = pool.iter().map(|path| path.len()).min()?;
    let short: Vec<&&String> = pool.iter().filter(|path| path.len() == shortest).collect();
    if short.len() == 1 {
        Some((*short[0]).clone())
    } else {
        None
    }
}

fn wiki_re() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"\[\[([^\[\]]+)\]\]").expect("static"))
}

fn md_link_re() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"\[([^\]]*)\]\(([^)]+)\)").expect("static"))
}

fn md_image_re() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"!\[[^\]]*\]\([^)]+\)").expect("static"))
}

fn wiki_embed_re() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"!\[\[[^\[\]]+\]\]").expect("static"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct ExtractCase {
        name: String,
        content: String,
        links: Vec<RawNoteLink>,
    }

    #[derive(Deserialize)]
    struct ResolveNote {
        #[serde(rename = "relativePath")]
        relative_path: String,
        title: String,
    }

    #[derive(Deserialize)]
    struct ResolveCase {
        name: String,
        #[serde(rename = "sourcePath")]
        source_path: String,
        #[serde(rename = "targetRef")]
        target_ref: String,
        notes: Vec<ResolveNote>,
        resolved: Option<String>,
    }

    #[derive(Deserialize)]
    struct Corpus {
        extract: Vec<ExtractCase>,
        resolve: Vec<ResolveCase>,
    }

    fn load_corpus() -> Corpus {
        let raw = include_str!(concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../src/features/library/fixtures/note-links-corpus.json"
        ));
        serde_json::from_str(raw).expect("corpus json")
    }

    #[test]
    fn extract_matches_shared_corpus() {
        let corpus = load_corpus();
        assert!(!corpus.extract.is_empty());
        for case in corpus.extract {
            assert_eq!(
                extract_note_links(&case.content),
                case.links,
                "extract {}",
                case.name
            );
        }
    }

    #[test]
    fn resolve_matches_shared_corpus() {
        let corpus = load_corpus();
        assert!(!corpus.resolve.is_empty());
        for case in corpus.resolve {
            let notes = case
                .notes
                .iter()
                .map(|note| NoteLinkIdentity {
                    relative_path: note.relative_path.clone(),
                    title: note.title.clone(),
                })
                .collect::<Vec<_>>();
            assert_eq!(
                resolve_note_ref(&case.target_ref, &case.source_path, &notes),
                case.resolved,
                "resolve {}",
                case.name
            );
        }
    }
}
